pymc_marketing/mmm/components/saturation.py

```diff
@@ -0,0 +1,170 @@
+#   Copyright 2024 The PyMC Labs Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""Saturation transformations for the MMM model.
+
+Each of these transformations is a subclass of
+:class:`pymc_marketing.mmm.components.saturation.SaturationTransformation` and defines a function
+that takes media and return the saturated media. The parameters of the function
+are the parameters of the saturation transformation.
+
+Examples
+--------
+Create a new saturation transformation:
+
+.. code-block:: python
+
+    class InfiniteReturns(SaturationTransformation):
+        def function(self, x, b):
+            return b * x
+
+        default_priors = {"b": {"dist": "HalfNormal", "kwargs": {"sigma": 1}}}
+
+"""
+
+from pymc_marketing.mmm.components.base import Transformation
+from pymc_marketing.mmm.transformers import (
+    hill_saturation,
+    logistic_saturation,
+    michaelis_menten,
+    tanh_saturation,
+    tanh_saturation_baselined,
+)
+
+
+class SaturationTransformation(Transformation):
+    """Subclass for all saturation transformations.
+
+    In order to use a custom saturation transformation, subclass and define:
+
+    - `function`: function to take x to contributions
+    - `default_priors`: default distributions for each parameter in function
+
+    By subclassing from this method, lift test integration will come for free!
+
+    Examples
+    ----------
+    Make a non-saturating saturation transformation
+
+    .. code-block:: python
+
+        def infinite_returns(x, b):
+            return b * x
+
+        class InfiniteReturns(SaturationTransformation):
+            function = infinite_returns
+            default_priors = {"b": {"dist": "HalfNormal", "kwargs": {"sigma": 1}}}
+
+    """
+
+    prefix: str = "saturation"
+
+
+class LogisticSaturation(SaturationTransformation):
+    """Wrapper around logistic saturation function.
+
+    For more information, see :func:`pymc_marketing.mmm.transformers.logistic_saturation`.
+
+    """
+
+    def function(self, x, lam, beta):
+        return beta * logistic_saturation(x, lam)
```

Comment:
Why are you adding a `beta` parameter? Will we have another regression coefficient in the PyMC Model? Will this work if we do the `LogisticSaturation` and then the `AdstockTransformation`?

Reply:
This is related to the saturation vs contribution function naming. 

This function output was scaled by beta in the larger MMM and now put with transformation. 

Saturation functions like Menten wouldn't need that beta scaling when used in the model as the output is contribution already. The idea was to have all saturation input map to same output space, namely spend -> contribution. 

Another option would be to consolidate on spend -> sat spend and pull out included alpha in menten.

Not too sure how the ordering affects the need for this beta. Maybe you have some thoughts? I was thinking it wouldnt have an effect 😅
Saturation first then has the spend -> contribution -> effective contribution as oppose to spend -> effective spend -> contribution with current order. 

@cetagostini Might have more intuition or opinion

Reply:
Thanks for the explanation! I think there is no right or wrong as it is a choice in parametrization. Still:

> Not too sure how the ordering affects the need for this beta. Maybe you have some thoughts? I was thinking it wouldnt have an effect t 😅

I think is worth checking so that we make sure our intuition is right :)

Reply:
The majority of our saturation functions are currently limited to saturating at 1. We can find double validate this in the [documentation](https://www.pymc-marketing.io/en/stable/api/generated/pymc_marketing.mmm.transformers.tanh_saturation_baselined.html) or by using our new interface. In any saturation function where the maximum response is cap to 1, the additional factor (Scaler) used for multiplication becomes its asymptote, representing the maximum value that the function saturation can reach.

Currently, `Menten` and `Tanh Baseline` are the only ones that have this value in their parameterization from the start. **We have also included the Hill function**, which has its scale incorporated into its parameterization. Only `Logistic` and `Tanh` had the escalator outside their parameterization, but it's easier to add it within the parameterization to avoid confusion.

Many users have interpreted `beta_channel` as a "coefficient," as if it were a pure linear regression. However, in our model, we don't have a coefficient; **the coefficient is the nonlinear transformation itself**, which depends on several factors.

Here is an [example](https://github.com/pymc-labs/pymc-marketing/discussions/617#discussioncomment-9244139) of that, along with the explanation I provided about it.

We believe it's better to make everything consistent, and avoid the confusion but what do you think?

cc: @wd60622

Reply:
Ok! Let's go for it an iterate through user feedback.